### ISSUE TITLE
Error during fisher exact test for ARAX_query.py 6237

### ISSUE THREAD

**Issue by amykglen** [RTXteam/RTX#837]
When I run `python ARAX_query.py 6237` locally (in `master`, with the latest changes), I'm getting this error:

`ERROR: An uncaught error occurred: A process in the process pool was terminated abruptly while the future was running or pending.: ['Traceback (most recent call last):\n', '  File "ARAX_query.py", line 501, in executeProcessingPlan\n    result = overlay.apply(message, action[\'parameters\'], response=response)\n', '  File "/Users/aglen/translator/RTX_new/RTX/code/ARAX/ARAXQuery/ARAX_overlay.py", line 142, in apply\n    getattr(self, \'_\' + self.__class__.__name__ + \'__\' + parameters[\'action\'])()  # thank you https://stackoverflow.com/questions/11649848/call-methods-by-string\n', '  File "/Users/aglen/translator/RTX_new/RTX/code/ARAX/ARAXQuery/ARAX_overlay.py", line 573, in __fisher_exact_test\n    response = FTEST.fisher_exact_test()\n', '  File "/Users/aglen/translator/RTX_new/RTX/code/ARAX/ARAXQuery/Overlay/fisher_exact_test.py", line 303, in fisher_exact_test\n    target_count_res = list(executor.map(self.query_adjacent_nodes_parallel, parament_list))\n', '  File "/Users/aglen/.pyenv/versions/3.7.3/lib/python3.7/concurrent/futures/process.py", line 476, in _chain_from_iterable_of_lists\n    for element in iterable:\n', '  File "/Users/aglen/.pyenv/versions/3.7.3/lib/python3.7/concurrent/futures/_base.py", line 586, in result_iterator\n    yield fs.pop().result()\n', '  File "/Users/aglen/.pyenv/versions/3.7.3/lib/python3.7/concurrent/futures/_base.py", line 432, in result\n    return self.__get_result()\n', '  File "/Users/aglen/.pyenv/versions/3.7.3/lib/python3.7/concurrent/futures/_base.py", line 384, in __get_result\n    raise self._exception\n', 'concurrent.futures.process.BrokenProcessPool: A process in the process pool was terminated abruptly while the future was running or pending.\n']`

Here's the full log:
```
(arax-env) amys-macbook:ARAXQuery aglen$ python ARAX_query.py 6237
Response:
  status: ERROR
  n_errors: 1  n_warnings: 0  n_messages: 208
  error_code: UncaughtARAXiError   message: An uncaught error occurred: A process in the process pool was terminated abruptly while the future was running or pending.: ['Traceback (most recent call last):\n', '  File "ARAX_query.py", line 501, in executeProcessingPlan\n    result = overlay.apply(message, action[\'parameters\'], response=response)\n', '  File "/Users/aglen/translator/RTX_new/RTX/code/ARAX/ARAXQuery/ARAX_overlay.py", line 142, in apply\n    getattr(self, \'_\' + self.__class__.__name__ + \'__\' + parameters[\'action\'])()  # thank you https://stackoverflow.com/questions/11649848/call-methods-by-string\n', '  File "/Users/aglen/translator/RTX_new/RTX/code/ARAX/ARAXQuery/ARAX_overlay.py", line 573, in __fisher_exact_test\n    response = FTEST.fisher_exact_test()\n', '  File "/Users/aglen/translator/RTX_new/RTX/code/ARAX/ARAXQuery/Overlay/fisher_exact_test.py", line 303, in fisher_exact_test\n    target_count_res = list(executor.map(self.query_adjacent_nodes_parallel, parament_list))\n', '  File "/Users/aglen/.pyenv/versions/3.7.3/lib/python3.7/concurrent/futures/process.py", line 476, in _chain_from_iterable_of_lists\n    for element in iterable:\n', '  File "/Users/aglen/.pyenv/versions/3.7.3/lib/python3.7/concurrent/futures/_base.py", line 586, in result_iterator\n    yield fs.pop().result()\n', '  File "/Users/aglen/.pyenv/versions/3.7.3/lib/python3.7/concurrent/futures/_base.py", line 432, in result\n    return self.__get_result()\n', '  File "/Users/aglen/.pyenv/versions/3.7.3/lib/python3.7/concurrent/futures/_base.py", line 384, in __get_result\n    raise self._exception\n', 'concurrent.futures.process.BrokenProcessPool: A process in the process pool was terminated abruptly while the future was running or pending.\n']
  - 2020-06-17 17:43:00.866238 INFO: ARAXQuery launching on incoming Message
  - 2020-06-17 17:43:00.866262 INFO: Examine input query for needed information for dispatch
  - 2020-06-17 17:43:00.866274 INFO: Found input processing plan. Sending to the ProcessingPlanExecutor
  - 2020-06-17 17:43:00.866281 DEBUG: Entering executeProcessingPlan
  - 2020-06-17 17:43:00.966586 DEBUG: No starting messages were referenced. Will start with a blank template Message
  - 2020-06-17 17:43:00.967270 DEBUG: Found processing_actions
  - 2020-06-17 17:43:00.967293 INFO: Parsing input actions list
  - 2020-06-17 17:43:00.967304 DEBUG: Parsing action: create_message
  - 2020-06-17 17:43:00.967795 DEBUG: Parsing action: add_qnode(id=n00, curie=CHEMBL.COMPOUND:CHEMBL521, type=chemical_substance)
  - 2020-06-17 17:43:00.968330 DEBUG: Parsing action: add_qnode(id=n01, is_set=true, type=protein)
  - 2020-06-17 17:43:00.968371 DEBUG: Parsing action: add_qedge(id=e00, source_id=n00, target_id=n01)
  - 2020-06-17 17:43:00.968404 DEBUG: Parsing action: expand(edge_id=e00, kp=ARAX/KG1)
  - 2020-06-17 17:43:00.968431 DEBUG: Parsing action: overlay(action=fisher_exact_test, source_qnode_id=n00, target_qnode_id=n01, virtual_relation_label=FET1)
  - 2020-06-17 17:43:00.968465 DEBUG: Parsing action: filter_kg(action=remove_edges_by_attribute, edge_attribute=fisher_exact_test_p-value, direction=above, threshold=0.01, remove_connected_nodes=t, qnode_id=n01)
  - 2020-06-17 17:43:00.968508 DEBUG: Parsing action: add_qnode(type=biological_process, is_set=true, id=n02)
  - 2020-06-17 17:43:00.968538 DEBUG: Parsing action: add_qedge(source_id=n01, target_id=n02, id=e01)
  - 2020-06-17 17:43:00.968567 DEBUG: Parsing action: expand(edge_id=e01, kp=ARAX/KG1)
  - 2020-06-17 17:43:00.968592 DEBUG: Parsing action: overlay(action=fisher_exact_test, source_qnode_id=n01, target_qnode_id=n02, virtual_relation_label=FET2)
  - 2020-06-17 17:43:00.968634 DEBUG: Parsing action: filter_kg(action=remove_edges_by_attribute, edge_attribute=fisher_exact_test_p-value, direction=above, threshold=0.01, remove_connected_nodes=t, qnode_id=n02)
  - 2020-06-17 17:43:00.968677 DEBUG: Parsing action: add_qnode(type=protein, is_set=true, id=n03)
  - 2020-06-17 17:43:00.968707 DEBUG: Parsing action: add_qedge(source_id=n02, target_id=n03, id=e02)
  - 2020-06-17 17:43:00.968739 DEBUG: Parsing action: expand(edge_id=e02, kp=ARAX/KG1)
  - 2020-06-17 17:43:00.968765 DEBUG: Parsing action: overlay(action=fisher_exact_test, source_qnode_id=n02, target_qnode_id=n03, virtual_relation_label=FET3)
  - 2020-06-17 17:43:00.968802 DEBUG: Parsing action: filter_kg(action=remove_edges_by_attribute, edge_attribute=fisher_exact_test_p-value, direction=above, threshold=0.01, remove_connected_nodes=t, qnode_id=n03)
  - 2020-06-17 17:43:00.968845 DEBUG: Parsing action: add_qnode(type=disease, id=n04)
  - 2020-06-17 17:43:00.968870 DEBUG: Parsing action: add_qedge(source_id=n03, target_id=n04, id=e03)
  - 2020-06-17 17:43:00.968903 DEBUG: Parsing action: expand(edge_id=e03, kp=ARAX/KG1)
  - 2020-06-17 17:43:00.968928 DEBUG: Parsing action: overlay(action=fisher_exact_test, source_qnode_id=n03, target_qnode_id=n04, virtual_relation_label=FET4)
  - 2020-06-17 17:43:00.968961 DEBUG: Parsing action: resultify()
  - 2020-06-17 17:43:00.968982 DEBUG: Parsing action: return(message=false, store=true)
  - 2020-06-17 17:43:00.983260 INFO: Processing action 'create_message' with parameters None
  - 2020-06-17 17:43:00.983314 INFO: Creating an empty template ARAX Message
  - 2020-06-17 17:43:00.984091 INFO: Processing action 'add_qnode' with parameters {'id': 'n00', 'curie': 'CHEMBL.COMPOUND:CHEMBL521', 'type': 'chemical_substance'}
  - 2020-06-17 17:43:00.984139 INFO: Adding a QueryNode to Message with parameters {'id': 'n00', 'curie': 'CHEMBL.COMPOUND:CHEMBL521', 'name': None, 'type': 'chemical_substance', 'is_set': None}
  - 2020-06-17 17:43:00.989946 DEBUG: Looking up CURIE CHEMBL.COMPOUND:CHEMBL521 in KgNodeIndex
  - 2020-06-17 17:43:00.998303 INFO: Processing action 'add_qnode' with parameters {'id': 'n01', 'is_set': 'true', 'type': 'protein'}
  - 2020-06-17 17:43:00.998383 INFO: Adding a QueryNode to Message with parameters {'id': 'n01', 'curie': None, 'name': None, 'type': 'protein', 'is_set': 'true'}
  - 2020-06-17 17:43:00.998977 INFO: Processing action 'add_qedge' with parameters {'id': 'e00', 'source_id': 'n00', 'target_id': 'n01'}
  - 2020-06-17 17:43:00.999030 INFO: Adding a QueryEdge to Message with parameters {'id': 'e00', 'source_id': 'n00', 'target_id': 'n01', 'type': None}
  - 2020-06-17 17:43:00.999095 INFO: Processing action 'expand' with parameters {'edge_id': 'e00', 'kp': 'ARAX/KG1'}
  - 2020-06-17 17:43:00.999136 DEBUG: Applying Expand to Message with parameters {'edge_id': 'e00', 'node_id': None, 'kp': 'ARAX/KG1', 'enforce_directionality': False, 'use_synonyms': True, 'synonym_handling': 'map_back', 'continue_if_no_results': False}
  - 2020-06-17 17:43:00.999421 DEBUG: Query graph for this Expand() call is: {'nodes': [{'id': 'n00', 'curie': 'CHEMBL.COMPOUND:CHEMBL521', 'type': 'chemical_substance', 'is_set': None}, {'id': 'n01', 'curie': None, 'type': 'protein', 'is_set': True}], 'edges': [{'id': 'e00', 'type': None, 'relation': None, 'source_id': 'n00', 'target_id': 'n01', 'negated': None}]}
  - 2020-06-17 17:43:00.999466 INFO: Expanding edge e00 using ARAX/KG1
  - 2020-06-17 17:43:01.003133 DEBUG: Looking for query nodes to use curie synonyms for
  - 2020-06-17 17:43:01.009339 DEBUG: Could not find any synonyms for curie CHEMBL.COMPOUND:CHEMBL521
  - 2020-06-17 17:43:01.009413 DEBUG: Generating cypher for edge e00 query graph
  - 2020-06-17 17:43:01.009474 INFO: Sending cypher query for edge e00 to KG1 neo4j
  - 2020-06-17 17:43:01.821799 INFO: Query for edge e00 returned results (e00: 56, n00: 1, n01: 28)
  - 2020-06-17 17:43:01.821830 DEBUG: Processing query results for edge e00
  - 2020-06-17 17:43:01.824114 DEBUG: Merging answer into Message.KnowledgeGraph
  - 2020-06-17 17:43:01.824211 DEBUG: Pruning any paths that are now dead ends
  - 2020-06-17 17:43:01.824646 INFO: After Expand, Message.KnowledgeGraph has 29 nodes and 56 edges
  - 2020-06-17 17:43:01.824676 INFO: Processing action 'overlay' with parameters {'action': 'fisher_exact_test', 'source_qnode_id': 'n00', 'target_qnode_id': 'n01', 'virtual_relation_label': 'FET1'}
  - 2020-06-17 17:43:01.840101 INFO: Performing Fisher's Exact Test to add p-value to edge attribute of virtual edge
  - 2020-06-17 17:43:01.840631 DEBUG: 1 source node with qnode id n00 and node type chemical_substance was found in message KG and used to calculate Fisher's Exact Test
  - 2020-06-17 17:43:01.840645 DEBUG: 28 target nodes with qnode id n01 and node type protein was found in message KG and used to calculate Fisher's Exact Test
  - 2020-06-17 17:43:01.840652 DEBUG: ARAX/KG1 was used to calculate total adjacent nodes in Fisher's Exact Test
  - 2020-06-17 17:43:09.582064 DEBUG: ARAX/KG1 and cypher query were used to calculate total number of node with the same type of source node in Fisher's Exact Test
  - 2020-06-17 17:43:09.582133 DEBUG: Total 2226 nodes with node type chemical_substance was found in ARAX/KG1
  - 2020-06-17 17:43:09.582147 DEBUG: Computing Fisher's Exact Test P-value
  - 2020-06-17 17:43:09.678870 DEBUG: Adding virtual edge with FET result to message KG
  - 2020-06-17 17:43:09.680181 DEBUG: 28 new virtual edges were added to message KG
  - 2020-06-17 17:43:09.680193 DEBUG: Adding virtual edge to message QG
  - 2020-06-17 17:43:09.680319 DEBUG: One virtual edge was added to message QG
  - 2020-06-17 17:43:09.680979 DEBUG: Applying Overlay to Message with parameters {'action': 'fisher_exact_test', 'source_qnode_id': 'n00', 'target_qnode_id': 'n01', 'virtual_relation_label': 'FET1'}
  - 2020-06-17 17:43:09.683444 DEBUG: Query graph is {'edges': [{'id': 'e00',
            'negated': None,
            'relation': None,
            'source_id': 'n00',
            'target_id': 'n01',
            'type': None},
           {'id': 'FET1',
            'negated': None,
            'relation': 'FET1',
            'source_id': 'n00',
            'target_id': 'n01',
            'type': 'has_fisher_exact_test_p-value_with'}],
 'nodes': [{'curie': 'CHEMBL.COMPOUND:CHEMBL521',
            'id': 'n00',
            'is_set': None,
            'type': 'chemical_substance'},
           {'curie': None, 'id': 'n01', 'is_set': True, 'type': 'protein'}]}
  - 2020-06-17 17:43:09.683500 DEBUG: Number of nodes in KG is 29
  - 2020-06-17 17:43:09.683733 DEBUG: Number of nodes in KG by type is Counter({'protein': 28, 'chemical_substance': 1})
  - 2020-06-17 17:43:09.683753 DEBUG: Number of edges in KG is 84
  - 2020-06-17 17:43:09.683881 DEBUG: Number of edges in KG by type is Counter({'physically_interacts_with': 56, 'has_fisher_exact_test_p-value_with': 28})
  - 2020-06-17 17:43:09.683948 DEBUG: Number of edges in KG with attributes is 70
  - 2020-06-17 17:43:09.684123 DEBUG: Number of edges in KG by attribute Counter({'probability': 42, 'fisher_exact_test_p-value': 28})
  - 2020-06-17 17:43:09.684181 INFO: Processing action 'filter_kg' with parameters {'action': 'remove_edges_by_attribute', 'edge_attribute': 'fisher_exact_test_p-value', 'direction': 'above', 'threshold': '0.01', 'remove_connected_nodes': 't', 'qnode_id': 'n01'}
  - 2020-06-17 17:43:09.690610 DEBUG: Removing Edges
  - 2020-06-17 17:43:09.690634 INFO: Removing edges from the knowledge graph with the specified attribute values
  - 2020-06-17 17:43:09.690864 DEBUG: Removing Nodes
  - 2020-06-17 17:43:09.690873 INFO: Removing connected nodes and their edges from the knowledge graph
  - 2020-06-17 17:43:09.691306 INFO: Edges successfully removed
  - 2020-06-17 17:43:09.691346 DEBUG: Applying Overlay to Message with parameters {'action': 'remove_edges_by_attribute', 'edge_attribute': 'fisher_exact_test_p-value', 'direction': 'above', 'threshold': 0.01, 'remove_connected_nodes': True, 'qnode_id': 'n01'}
  - 2020-06-17 17:43:09.692049 DEBUG: Query graph is {'edges': [{'id': 'e00',
            'negated': None,
            'relation': None,
            'source_id': 'n00',
            'target_id': 'n01',
            'type': None},
           {'id': 'FET1',
            'negated': None,
            'relation': 'FET1',
            'source_id': 'n00',
            'target_id': 'n01',
            'type': 'has_fisher_exact_test_p-value_with'}],
 'nodes': [{'curie': 'CHEMBL.COMPOUND:CHEMBL521',
            'id': 'n00',
            'is_set': None,
            'type': 'chemical_substance'},
           {'curie': None, 'id': 'n01', 'is_set': True, 'type': 'protein'}]}
  - 2020-06-17 17:43:09.692070 DEBUG: Number of nodes in KG is 2
  - 2020-06-17 17:43:09.692119 DEBUG: Number of nodes in KG by type is Counter({'chemical_substance': 1, 'protein': 1})
  - 2020-06-17 17:43:09.692130 DEBUG: Number of edges in KG is 3
  - 2020-06-17 17:43:09.692159 DEBUG: Number of edges in KG by type is Counter({'physically_interacts_with': 2, 'has_fisher_exact_test_p-value_with': 1})
  - 2020-06-17 17:43:09.692171 DEBUG: Number of edges in KG with attributes is 3
  - 2020-06-17 17:43:09.692199 DEBUG: Number of edges in KG by attribute Counter({'probability': 2, 'fisher_exact_test_p-value': 1})
  - 2020-06-17 17:43:09.692225 INFO: Processing action 'add_qnode' with parameters {'type': 'biological_process', 'is_set': 'true', 'id': 'n02'}
  - 2020-06-17 17:43:09.692274 INFO: Adding a QueryNode to Message with parameters {'id': 'n02', 'curie': None, 'name': None, 'type': 'biological_process', 'is_set': 'true'}
  - 2020-06-17 17:43:09.692946 INFO: Processing action 'add_qedge' with parameters {'source_id': 'n01', 'target_id': 'n02', 'id': 'e01'}
  - 2020-06-17 17:43:09.692992 INFO: Adding a QueryEdge to Message with parameters {'id': 'e01', 'source_id': 'n01', 'target_id': 'n02', 'type': None}
  - 2020-06-17 17:43:09.693026 INFO: Processing action 'expand' with parameters {'edge_id': 'e01', 'kp': 'ARAX/KG1'}
  - 2020-06-17 17:43:09.693078 DEBUG: Applying Expand to Message with parameters {'edge_id': 'e01', 'node_id': None, 'kp': 'ARAX/KG1', 'enforce_directionality': False, 'use_synonyms': True, 'synonym_handling': 'map_back', 'continue_if_no_results': False}
  - 2020-06-17 17:43:09.693344 DEBUG: Query graph for this Expand() call is: {'nodes': [{'id': 'n01', 'curie': None, 'type': 'protein', 'is_set': True}, {'id': 'n02', 'curie': None, 'type': 'biological_process', 'is_set': True}], 'edges': [{'id': 'e01', 'type': None, 'relation': None, 'source_id': 'n01', 'target_id': 'n02', 'negated': None}]}
  - 2020-06-17 17:43:09.693382 INFO: Expanding edge e01 using ARAX/KG1
  - 2020-06-17 17:43:09.693494 DEBUG: Looking for query nodes to use curie synonyms for
  - 2020-06-17 17:43:09.693512 DEBUG: Generating cypher for edge e01 query graph
  - 2020-06-17 17:43:09.693557 INFO: Sending cypher query for edge e01 to KG1 neo4j
  - 2020-06-17 17:43:10.301633 INFO: Query for edge e01 returned results (e01: 6, n01: 1, n02: 6)
  - 2020-06-17 17:43:10.301678 DEBUG: Processing query results for edge e01
  - 2020-06-17 17:43:10.302184 DEBUG: Merging answer into Message.KnowledgeGraph
  - 2020-06-17 17:43:10.302216 DEBUG: Pruning any paths that are now dead ends
  - 2020-06-17 17:43:10.302358 INFO: After Expand, Message.KnowledgeGraph has 8 nodes and 9 edges
  - 2020-06-17 17:43:10.302385 INFO: Processing action 'overlay' with parameters {'action': 'fisher_exact_test', 'source_qnode_id': 'n01', 'target_qnode_id': 'n02', 'virtual_relation_label': 'FET2'}
  - 2020-06-17 17:43:10.302465 INFO: Performing Fisher's Exact Test to add p-value to edge attribute of virtual edge
  - 2020-06-17 17:43:10.302565 DEBUG: 1 source node with qnode id n01 and node type protein was found in message KG and used to calculate Fisher's Exact Test
  - 2020-06-17 17:43:10.302575 DEBUG: 6 target nodes with qnode id n02 and node type biological_process was found in message KG and used to calculate Fisher's Exact Test
  - 2020-06-17 17:43:10.302581 DEBUG: ARAX/KG1 was used to calculate total adjacent nodes in Fisher's Exact Test
  - 2020-06-17 17:43:13.002912 DEBUG: ARAX/KG1 and cypher query were used to calculate total number of node with the same type of source node in Fisher's Exact Test
  - 2020-06-17 17:43:13.002987 DEBUG: Total 20514 nodes with node type protein was found in ARAX/KG1
  - 2020-06-17 17:43:13.002999 DEBUG: Computing Fisher's Exact Test P-value
  - 2020-06-17 17:43:13.041390 DEBUG: Adding virtual edge with FET result to message KG
  - 2020-06-17 17:43:13.042082 DEBUG: 6 new virtual edges were added to message KG
  - 2020-06-17 17:43:13.042092 DEBUG: Adding virtual edge to message QG
  - 2020-06-17 17:43:13.042175 DEBUG: One virtual edge was added to message QG
  - 2020-06-17 17:43:13.042413 DEBUG: Applying Overlay to Message with parameters {'action': 'fisher_exact_test', 'source_qnode_id': 'n01', 'target_qnode_id': 'n02', 'virtual_relation_label': 'FET2'}
  - 2020-06-17 17:43:13.043636 DEBUG: Query graph is {'edges': [{'id': 'e00',
            'negated': None,
            'relation': None,
            'source_id': 'n00',
            'target_id': 'n01',
            'type': None},
           {'id': 'FET1',
            'negated': None,
            'relation': 'FET1',
            'source_id': 'n00',
            'target_id': 'n01',
            'type': 'has_fisher_exact_test_p-value_with'},
           {'id': 'e01',
            'negated': None,
            'relation': None,
            'source_id': 'n01',
            'target_id': 'n02',
            'type': None},
           {'id': 'FET2',
            'negated': None,
            'relation': 'FET2',
            'source_id': 'n01',
            'target_id': 'n02',
            'type': 'has_fisher_exact_test_p-value_with'}],
 'nodes': [{'curie': 'CHEMBL.COMPOUND:CHEMBL521',
            'id': 'n00',
            'is_set': None,
            'type': 'chemical_substance'},
           {'curie': None, 'id': 'n01', 'is_set': True, 'type': 'protein'},
           {'curie': None,
            'id': 'n02',
            'is_set': True,
            'type': 'biological_process'}]}
  - 2020-06-17 17:43:13.043655 DEBUG: Number of nodes in KG is 8
  - 2020-06-17 17:43:13.043769 DEBUG: Number of nodes in KG by type is Counter({'biological_process': 6, 'chemical_substance': 1, 'protein': 1})
  - 2020-06-17 17:43:13.043780 DEBUG: Number of edges in KG is 15
  - 2020-06-17 17:43:13.043812 DEBUG: Number of edges in KG by type is Counter({'has_fisher_exact_test_p-value_with': 7, 'involved_in': 6, 'physically_interacts_with': 2})
  - 2020-06-17 17:43:13.043828 DEBUG: Number of edges in KG with attributes is 9
  - 2020-06-17 17:43:13.043862 DEBUG: Number of edges in KG by attribute Counter({'fisher_exact_test_p-value': 7, 'probability': 2})
  - 2020-06-17 17:43:13.043989 INFO: Processing action 'filter_kg' with parameters {'action': 'remove_edges_by_attribute', 'edge_attribute': 'fisher_exact_test_p-value', 'direction': 'above', 'threshold': '0.01', 'remove_connected_nodes': 't', 'qnode_id': 'n02'}
  - 2020-06-17 17:43:13.044139 DEBUG: Removing Edges
  - 2020-06-17 17:43:13.044149 INFO: Removing edges from the knowledge graph with the specified attribute values
  - 2020-06-17 17:43:13.044211 DEBUG: Removing Nodes
  - 2020-06-17 17:43:13.044219 INFO: Removing connected nodes and their edges from the knowledge graph
  - 2020-06-17 17:43:13.044271 INFO: Edges successfully removed
  - 2020-06-17 17:43:13.044289 DEBUG: Applying Overlay to Message with parameters {'action': 'remove_edges_by_attribute', 'edge_attribute': 'fisher_exact_test_p-value', 'direction': 'above', 'threshold': 0.01, 'remove_connected_nodes': True, 'qnode_id': 'n02'}
  - 2020-06-17 17:43:13.045120 DEBUG: Query graph is {'edges': [{'id': 'e00',
            'negated': None,
            'relation': None,
            'source_id': 'n00',
            'target_id': 'n01',
            'type': None},
           {'id': 'FET1',
            'negated': None,
            'relation': 'FET1',
            'source_id': 'n00',
            'target_id': 'n01',
            'type': 'has_fisher_exact_test_p-value_with'},
           {'id': 'e01',
            'negated': None,
            'relation': None,
            'source_id': 'n01',
            'target_id': 'n02',
            'type': None},
           {'id': 'FET2',
            'negated': None,
            'relation': 'FET2',
            'source_id': 'n01',
            'target_id': 'n02',
            'type': 'has_fisher_exact_test_p-value_with'}],
 'nodes': [{'curie': 'CHEMBL.COMPOUND:CHEMBL521',
            'id': 'n00',
            'is_set': None,
            'type': 'chemical_substance'},
           {'curie': None, 'id': 'n01', 'is_set': True, 'type': 'protein'},
           {'curie': None,
            'id': 'n02',
            'is_set': True,
            'type': 'biological_process'}]}
  - 2020-06-17 17:43:13.045135 DEBUG: Number of nodes in KG is 7
  - 2020-06-17 17:43:13.045168 DEBUG: Number of nodes in KG by type is Counter({'biological_process': 5, 'chemical_substance': 1, 'protein': 1})
  - 2020-06-17 17:43:13.045176 DEBUG: Number of edges in KG is 13
  - 2020-06-17 17:43:13.045202 DEBUG: Number of edges in KG by type is Counter({'has_fisher_exact_test_p-value_with': 6, 'involved_in': 5, 'physically_interacts_with': 2})
  - 2020-06-17 17:43:13.045213 DEBUG: Number of edges in KG with attributes is 8
  - 2020-06-17 17:43:13.045240 DEBUG: Number of edges in KG by attribute Counter({'fisher_exact_test_p-value': 6, 'probability': 2})
  - 2020-06-17 17:43:13.045264 INFO: Processing action 'add_qnode' with parameters {'type': 'protein', 'is_set': 'true', 'id': 'n03'}
  - 2020-06-17 17:43:13.045348 INFO: Adding a QueryNode to Message with parameters {'id': 'n03', 'curie': None, 'name': None, 'type': 'protein', 'is_set': 'true'}
  - 2020-06-17 17:43:13.045938 INFO: Processing action 'add_qedge' with parameters {'source_id': 'n02', 'target_id': 'n03', 'id': 'e02'}
  - 2020-06-17 17:43:13.045968 INFO: Adding a QueryEdge to Message with parameters {'id': 'e02', 'source_id': 'n02', 'target_id': 'n03', 'type': None}
  - 2020-06-17 17:43:13.045994 INFO: Processing action 'expand' with parameters {'edge_id': 'e02', 'kp': 'ARAX/KG1'}
  - 2020-06-17 17:43:13.046032 DEBUG: Applying Expand to Message with parameters {'edge_id': 'e02', 'node_id': None, 'kp': 'ARAX/KG1', 'enforce_directionality': False, 'use_synonyms': True, 'synonym_handling': 'map_back', 'continue_if_no_results': False}
  - 2020-06-17 17:43:13.046215 DEBUG: Query graph for this Expand() call is: {'nodes': [{'id': 'n02', 'curie': None, 'type': 'biological_process', 'is_set': True}, {'id': 'n03', 'curie': None, 'type': 'protein', 'is_set': True}], 'edges': [{'id': 'e02', 'type': None, 'relation': None, 'source_id': 'n02', 'target_id': 'n03', 'negated': None}]}
  - 2020-06-17 17:43:13.046244 INFO: Expanding edge e02 using ARAX/KG1
  - 2020-06-17 17:43:13.046330 DEBUG: Looking for query nodes to use curie synonyms for
  - 2020-06-17 17:43:13.046344 DEBUG: Generating cypher for edge e02 query graph
  - 2020-06-17 17:43:13.046380 INFO: Sending cypher query for edge e02 to KG1 neo4j
  - 2020-06-17 17:43:14.142707 INFO: Query for edge e02 returned results (e02: 158, n02: 5, n03: 154)
  - 2020-06-17 17:43:14.142742 DEBUG: Processing query results for edge e02
  - 2020-06-17 17:43:14.148534 DEBUG: Merging answer into Message.KnowledgeGraph
  - 2020-06-17 17:43:14.148772 DEBUG: Pruning any paths that are now dead ends
  - 2020-06-17 17:43:14.150175 INFO: After Expand, Message.KnowledgeGraph has 160 nodes and 166 edges
  - 2020-06-17 17:43:14.150225 INFO: Processing action 'overlay' with parameters {'action': 'fisher_exact_test', 'source_qnode_id': 'n02', 'target_qnode_id': 'n03', 'virtual_relation_label': 'FET3'}
  - 2020-06-17 17:43:14.150447 INFO: Performing Fisher's Exact Test to add p-value to edge attribute of virtual edge
  - 2020-06-17 17:43:14.151299 DEBUG: 5 source nodes with qnode id n02 and node type biological_process was found in message KG and used to calculate Fisher's Exact Test
  - 2020-06-17 17:43:14.151309 DEBUG: 154 target nodes with qnode id n03 and node type protein was found in message KG and used to calculate Fisher's Exact Test
  - 2020-06-17 17:43:14.151315 DEBUG: ARAX/KG1 was used to calculate total adjacent nodes in Fisher's Exact Test
  - 2020-06-17 17:43:57.282878 DEBUG: ARAX/KG1 and cypher query were used to calculate total number of node with the same type of source node in Fisher's Exact Test
  - 2020-06-17 17:43:57.282960 DEBUG: Total 30716 nodes with node type biological_process was found in ARAX/KG1
  - 2020-06-17 17:43:57.282973 DEBUG: Computing Fisher's Exact Test P-value
  - 2020-06-17 17:43:58.097236 DEBUG: Adding virtual edge with FET result to message KG
  - 2020-06-17 17:43:58.102547 DEBUG: 158 new virtual edges were added to message KG
  - 2020-06-17 17:43:58.102560 DEBUG: Adding virtual edge to message QG
  - 2020-06-17 17:43:58.102615 DEBUG: One virtual edge was added to message QG
  - 2020-06-17 17:43:58.103505 DEBUG: Applying Overlay to Message with parameters {'action': 'fisher_exact_test', 'source_qnode_id': 'n02', 'target_qnode_id': 'n03', 'virtual_relation_label': 'FET3'}
  - 2020-06-17 17:43:58.104907 DEBUG: Query graph is {'edges': [{'id': 'e00',
            'negated': None,
            'relation': None,
            'source_id': 'n00',
            'target_id': 'n01',
            'type': None},
           {'id': 'FET1',
            'negated': None,
            'relation': 'FET1',
            'source_id': 'n00',
            'target_id': 'n01',
            'type': 'has_fisher_exact_test_p-value_with'},
           {'id': 'e01',
            'negated': None,
            'relation': None,
            'source_id': 'n01',
            'target_id': 'n02',
            'type': None},
           {'id': 'FET2',
            'negated': None,
            'relation': 'FET2',
            'source_id': 'n01',
            'target_id': 'n02',
            'type': 'has_fisher_exact_test_p-value_with'},
           {'id': 'e02',
            'negated': None,
            'relation': None,
            'source_id': 'n02',
            'target_id': 'n03',
            'type': None},
           {'id': 'FET3',
            'negated': None,
            'relation': 'FET3',
            'source_id': 'n02',
            'target_id': 'n03',
            'type': 'has_fisher_exact_test_p-value_with'}],
 'nodes': [{'curie': 'CHEMBL.COMPOUND:CHEMBL521',
            'id': 'n00',
            'is_set': None,
            'type': 'chemical_substance'},
           {'curie': None, 'id': 'n01', 'is_set': True, 'type': 'protein'},
           {'curie': None,
            'id': 'n02',
            'is_set': True,
            'type': 'biological_process'},
           {'curie': None, 'id': 'n03', 'is_set': True, 'type': 'protein'}]}
  - 2020-06-17 17:43:58.104927 DEBUG: Number of nodes in KG is 160
  - 2020-06-17 17:43:58.105193 DEBUG: Number of nodes in KG by type is Counter({'protein': 154, 'biological_process': 5, 'chemical_substance': 1})
  - 2020-06-17 17:43:58.105204 DEBUG: Number of edges in KG is 324
  - 2020-06-17 17:43:58.105372 DEBUG: Number of edges in KG by type is Counter({'has_fisher_exact_test_p-value_with': 164, 'involved_in': 158, 'physically_interacts_with': 2})
  - 2020-06-17 17:43:58.105469 DEBUG: Number of edges in KG with attributes is 166
  - 2020-06-17 17:43:58.105677 DEBUG: Number of edges in KG by attribute Counter({'fisher_exact_test_p-value': 164, 'probability': 2})
  - 2020-06-17 17:43:58.105702 INFO: Processing action 'filter_kg' with parameters {'action': 'remove_edges_by_attribute', 'edge_attribute': 'fisher_exact_test_p-value', 'direction': 'above', 'threshold': '0.01', 'remove_connected_nodes': 't', 'qnode_id': 'n03'}
  - 2020-06-17 17:43:58.106420 DEBUG: Removing Edges
  - 2020-06-17 17:43:58.106437 INFO: Removing edges from the knowledge graph with the specified attribute values
  - 2020-06-17 17:43:58.106943 DEBUG: Removing Nodes
  - 2020-06-17 17:43:58.106952 INFO: Removing connected nodes and their edges from the knowledge graph
  - 2020-06-17 17:43:58.107481 INFO: Edges successfully removed
  - 2020-06-17 17:43:58.107512 DEBUG: Applying Overlay to Message with parameters {'action': 'remove_edges_by_attribute', 'edge_attribute': 'fisher_exact_test_p-value', 'direction': 'above', 'threshold': 0.01, 'remove_connected_nodes': True, 'qnode_id': 'n03'}
  - 2020-06-17 17:43:58.108668 DEBUG: Query graph is {'edges': [{'id': 'e00',
            'negated': None,
            'relation': None,
            'source_id': 'n00',
            'target_id': 'n01',
            'type': None},
           {'id': 'FET1',
            'negated': None,
            'relation': 'FET1',
            'source_id': 'n00',
            'target_id': 'n01',
            'type': 'has_fisher_exact_test_p-value_with'},
           {'id': 'e01',
            'negated': None,
            'relation': None,
            'source_id': 'n01',
            'target_id': 'n02',
            'type': None},
           {'id': 'FET2',
            'negated': None,
            'relation': 'FET2',
            'source_id': 'n01',
            'target_id': 'n02',
            'type': 'has_fisher_exact_test_p-value_with'},
           {'id': 'e02',
            'negated': None,
            'relation': None,
            'source_id': 'n02',
            'target_id': 'n03',
            'type': None},
           {'id': 'FET3',
            'negated': None,
            'relation': 'FET3',
            'source_id': 'n02',
            'target_id': 'n03',
            'type': 'has_fisher_exact_test_p-value_with'}],
 'nodes': [{'curie': 'CHEMBL.COMPOUND:CHEMBL521',
            'id': 'n00',
            'is_set': None,
            'type': 'chemical_substance'},
           {'curie': None, 'id': 'n01', 'is_set': True, 'type': 'protein'},
           {'curie': None,
            'id': 'n02',
            'is_set': True,
            'type': 'biological_process'},
           {'curie': None, 'id': 'n03', 'is_set': True, 'type': 'protein'}]}
  - 2020-06-17 17:43:58.108684 DEBUG: Number of nodes in KG is 135
  - 2020-06-17 17:43:58.108774 DEBUG: Number of nodes in KG by type is Counter({'protein': 129, 'biological_process': 5, 'chemical_substance': 1})
  - 2020-06-17 17:43:58.108784 DEBUG: Number of edges in KG is 274
  - 2020-06-17 17:43:58.108893 DEBUG: Number of edges in KG by type is Counter({'has_fisher_exact_test_p-value_with': 139, 'involved_in': 133, 'physically_interacts_with': 2})
  - 2020-06-17 17:43:58.108966 DEBUG: Number of edges in KG with attributes is 141
  - 2020-06-17 17:43:58.109140 DEBUG: Number of edges in KG by attribute Counter({'fisher_exact_test_p-value': 139, 'probability': 2})
  - 2020-06-17 17:43:58.109162 INFO: Processing action 'add_qnode' with parameters {'type': 'disease', 'id': 'n04'}
  - 2020-06-17 17:43:58.109246 INFO: Adding a QueryNode to Message with parameters {'id': 'n04', 'curie': None, 'name': None, 'type': 'disease', 'is_set': None}
  - 2020-06-17 17:43:58.110250 INFO: Processing action 'add_qedge' with parameters {'source_id': 'n03', 'target_id': 'n04', 'id': 'e03'}
  - 2020-06-17 17:43:58.110289 INFO: Adding a QueryEdge to Message with parameters {'id': 'e03', 'source_id': 'n03', 'target_id': 'n04', 'type': None}
  - 2020-06-17 17:43:58.110321 INFO: Processing action 'expand' with parameters {'edge_id': 'e03', 'kp': 'ARAX/KG1'}
  - 2020-06-17 17:43:58.110363 DEBUG: Applying Expand to Message with parameters {'edge_id': 'e03', 'node_id': None, 'kp': 'ARAX/KG1', 'enforce_directionality': False, 'use_synonyms': True, 'synonym_handling': 'map_back', 'continue_if_no_results': False}
  - 2020-06-17 17:43:58.110896 DEBUG: Query graph for this Expand() call is: {'nodes': [{'id': 'n03', 'curie': None, 'type': 'protein', 'is_set': True}, {'id': 'n04', 'curie': None, 'type': 'disease', 'is_set': None}], 'edges': [{'id': 'e03', 'type': None, 'relation': None, 'source_id': 'n03', 'target_id': 'n04', 'negated': None}]}
  - 2020-06-17 17:43:58.110925 INFO: Expanding edge e03 using ARAX/KG1
  - 2020-06-17 17:43:58.111021 DEBUG: Looking for query nodes to use curie synonyms for
  - 2020-06-17 17:43:58.111035 DEBUG: Generating cypher for edge e03 query graph
  - 2020-06-17 17:43:58.111099 INFO: Sending cypher query for edge e03 to KG1 neo4j
  - 2020-06-17 17:44:00.819453 INFO: Query for edge e03 returned results (e03: 1498, n03: 92, n04: 995)
  - 2020-06-17 17:44:00.819492 DEBUG: Processing query results for edge e03
  - 2020-06-17 17:44:00.869790 DEBUG: Merging answer into Message.KnowledgeGraph
  - 2020-06-17 17:44:00.872410 DEBUG: Pruning any paths that are now dead ends
  - 2020-06-17 17:44:00.896608 INFO: After Expand, Message.KnowledgeGraph has 1130 nodes and 1772 edges
  - 2020-06-17 17:44:00.897334 INFO: Processing action 'overlay' with parameters {'action': 'fisher_exact_test', 'source_qnode_id': 'n03', 'target_qnode_id': 'n04', 'virtual_relation_label': 'FET4'}
  - 2020-06-17 17:44:00.898834 INFO: Performing Fisher's Exact Test to add p-value to edge attribute of virtual edge
  - 2020-06-17 17:44:00.908509 DEBUG: 92 source nodes with qnode id n03 and node type protein was found in message KG and used to calculate Fisher's Exact Test
  - 2020-06-17 17:44:00.908537 DEBUG: 995 target nodes with qnode id n04 and node type disease was found in message KG and used to calculate Fisher's Exact Test
  - 2020-06-17 17:44:00.908544 DEBUG: ARAX/KG1 was used to calculate total adjacent nodes in Fisher's Exact Test
  - 2020-06-17 17:44:36.986836 ERROR: An uncaught error occurred: A process in the process pool was terminated abruptly while the future was running or pending.: ['Traceback (most recent call last):\n', '  File "ARAX_query.py", line 501, in executeProcessingPlan\n    result = overlay.apply(message, action[\'parameters\'], response=response)\n', '  File "/Users/aglen/translator/RTX_new/RTX/code/ARAX/ARAXQuery/ARAX_overlay.py", line 142, in apply\n    getattr(self, \'_\' + self.__class__.__name__ + \'__\' + parameters[\'action\'])()  # thank you https://stackoverflow.com/questions/11649848/call-methods-by-string\n', '  File "/Users/aglen/translator/RTX_new/RTX/code/ARAX/ARAXQuery/ARAX_overlay.py", line 573, in __fisher_exact_test\n    response = FTEST.fisher_exact_test()\n', '  File "/Users/aglen/translator/RTX_new/RTX/code/ARAX/ARAXQuery/Overlay/fisher_exact_test.py", line 303, in fisher_exact_test\n    target_count_res = list(executor.map(self.query_adjacent_nodes_parallel, parament_list))\n', '  File "/Users/aglen/.pyenv/versions/3.7.3/lib/python3.7/concurrent/futures/process.py", line 476, in _chain_from_iterable_of_lists\n    for element in iterable:\n', '  File "/Users/aglen/.pyenv/versions/3.7.3/lib/python3.7/concurrent/futures/_base.py", line 586, in result_iterator\n    yield fs.pop().result()\n', '  File "/Users/aglen/.pyenv/versions/3.7.3/lib/python3.7/concurrent/futures/_base.py", line 432, in result\n    return self.__get_result()\n', '  File "/Users/aglen/.pyenv/versions/3.7.3/lib/python3.7/concurrent/futures/_base.py", line 384, in __get_result\n    raise self._exception\n', 'concurrent.futures.process.BrokenProcessPool: A process in the process pool was terminated abruptly while the future was running or pending.\n']
```

**Comment by chunyuma** [comment]
@amykglen, I just ran it on `chunyu.rtx.ai` with the latest code without any problems. [Here](https://arax.rtx.ai/api/rtx/v1/message/62) is the result. Based on the log you posted, the error seems coming from the step of querying the adjacent nodes in parallel in the last hop. But it can successfully pass the first 3 hops. It's wired. Could you please try running it more one time?

**Comment by dkoslicki** [comment]
@chunyuma I see that error sometimes when the parallel pool workers have not finished executing before the script terminates (or moves on to some next step). You’re using a different kind of parallelization that I typically do (I use `multiprocessing`), so I don’t know the exact commands, but you may need to have something like a `executor.wait()`, or `.close()`, or something like that. I’ve also found that explicitly consuming the output sometimes also works (like `[x for x in executor.map(self.query_adjacent_nodes_parallel, parament_list)]` instead of `list(executor.map(self.query_adjacent_nodes_parallel, parament_list))`) as this guarantees that all workers will have returned their results (as list() and map() are lazy iterables that may/may not fully consume the results)

**Comment by chunyuma** [comment]
> I see that error sometimes when the parallel pool workers have not finished executing before the script terminates (or moves on to some next step). You’re using a different kind of parallelization than I typically do (I use multiprocessing), so I don’t know the exact commands, but you may need to have something like a executor.wait(), or .close(), or something like that.

@dkoslicki, I didn't know it. Great hint. I will try to check if `concurrent.futures` module has such method to solve this. BTW, `concurrent.futures` is a standard library module built into Python 3 and so it's more compatible with Python3.

>  I’ve also found that explicitly consuming the output sometimes also works (like [x for x in executor.map(self.query_adjacent_nodes_parallel, parament_list)] instead of list(executor.map(self.query_adjacent_nodes_parallel, parament_list))) as this guarantees that all workers will have returned their results (as list() and map() are lazy iterables that may/may not fully consume the results).

I can use `list comprehension` to replace `list()`. But one thing I worry about is that `list comprehensive` might make it a little bit slower as it's essentially a for loop.  There are some comparisons between `list comprehension` and `list()` online saying that `list()` might be faster than `list comprehension` as `list()` is written in C code. [Here](https://stackoverflow.com/questions/29356846/python-list-vs-list-comprehension-building-speed) is one of comparisons. But anyway I can check it.

**Comment by amykglen** [comment]
fyi, @chunyuma: I've tried running ARAX_query.py 6237 three times now, and all of them gave this error. but three-hop FET queries (like 6236) complete successfully for me.

**Comment by chunyuma** [comment]
thanks @amykglen, this might be caused by the reasons mentioned by @dkoslicki.  I will try to see if the methods suggested by @dkoslicki can solve it. But problem is that it is hard for me to test it because I ran it with no problem on ` chunyu.rtx.ai` and even on /test. Did you run it on macbook? Perhaps @amykglen I might need your help for testing once I modify the script.

**Comment by amykglen** [comment]
yep, running it on a mac. and no problem to help with testing - just let me know whenever you're ready.

**Comment by chunyuma** [comment]
@amykglen, I change the `concurrent.futures` to `multiprocessing` module to see if this can solve this error as I found the similar errors `A process in the process pool was terminated abruptly while the future was running or pending` reported online mostly come from `concurrent.futures`. Could you help me do a test? Thank you!

**Comment by amykglen** [comment]
just tried it out and it worked! (ARAX_query.py 6237 completed successfully with no errors.)

**Comment by chunyuma** [comment]
Awesome, thanks @amykglen

**Comment by dkoslicki** [comment]
Great to hear it's working!

 And just an FYI @chunyuma, while list comprehension is slower than list(), care must be taken when doing multithreaded/multitasking work as you want to make sure you actually consume everything (and not just creat a dynamic iterable).

Also, in general, be aware of premature optimization: the lists you're dealing with are rather small, so it's good to not worry too much about speed concerns for a particular data structures until you have hard evidence that you have a bottleneck.

**Comment by amykglen** [comment]
so this sounds crazy, but now when I run FET tests in ARAX_query.py (6237, 6235, etc.), Resultify seems to be being called a TON of times behind the scenes, even though there is of course only one `resultify` command in the DSL. this does not seem to be happening with queries that do not use FET.

to see what I'm talking about, locally (in `master`):
1. add `print("Resultify is being called")` as the first line of Resultify's `apply` method ([here](https://github.com/RTXteam/RTX/blob/5fcf8b17a1d4ba2b91af58f562b390cc34f6cb7a/code/ARAX/ARAXQuery/ARAX_resultify.py#L118))
2. run ARAX_query.py 6235 (which uses FET) - and you'll see something like 200 'Resultify is being called' lines printed
3. run ARAX_query.py 101 (which does not use FET) - and you'll see only 1 'Resultify is being called' line printed (as should be the case)

posting this here because I'm wondering if it's related to the changes made in this issue, but let me know if you'd like this written up in a separate issue.

**Comment by chunyuma** [comment]
@amykglen, the reason is because within FET, for each query node, it calls one DSL [here](https://github.com/RTXteam/RTX/blob/5fcf8b17a1d4ba2b91af58f562b390cc34f6cb7a/code/ARAX/ARAXQuery/Overlay/fisher_exact_test.py#L645-L664). So for 6235, it is a 2-hop query and it might contain around 200 query nodes.

**Comment by amykglen** [comment]
ohh, that makes sense. I see... forgot about how you call DSL from within DSL. all seems well then. thanks!

**Comment by amykglen** [comment]
I don't think there's really testing on production to do for this issue, since the error wasn't observed on the server - so I'll go ahead and close it.
